### PR TITLE
Feat: channel pins

### LINKED ISSUE #271: Channel Pins Support
Routes.channelPin // message.pin() / channel.pins.add() / channel.pins.remove()
Routes.channelPins // channel.pins.fetch()

Channel#pins should be a ProxyCache to Channel#messages with add() / remove() / fetch() methods. Message#pin() would internally call this.channel.pins.add(this.id). Could probably do the same for a Message#unpin()

### PR DESCRIPTION
**Pull request by Relms12345**
Fixes #271

### CODE REVIEW

File: src/lib/caching/structures/channels/GuildTextChannel.ts [hunk 1]
```diff
@@ -63,6 +70,9 @@ export abstract class GuildTextChannel extends GuildChannel {
 		super(client, data, guild);
 		this.messages = new MessageStore(client, this as unknown as GuildTextBasedChannel);
 		this.typing = new Typing(this);
+
+		const filterMessagePins = this.messages.filter(msg => msg.pinned).keys();
```
Comment (bdistin) [inline src/lib/caching/structures/channels/GuildTextChannel.ts:74]: At this point, the MessageStore will always be empty.

File: src/lib/caching/structures/guilds/Guild.ts [hunk 2]
```diff
@@ -496,6 +497,14 @@ export class Guild extends Structure {
 		return this.icon ? this.client.api.cdn.guildIcon(this.id, this.icon, options) : null;
 	}
 
+	/**
+	 * Fetches the webhooks the guild has
+	 * @since 0.0.4
+	 */
+	public async fetchWebhooks(options: RequestOptions = {}): Promise<APIWebhookData[]> {
```
Comment (bdistin) [inline src/lib/caching/structures/guilds/Guild.ts:504]: This is a separate feature.

File: src/lib/caching/structures/messages/Message.ts [hunk 3]
```diff
@@ -278,6 +279,22 @@ export class Message extends WebhookMessage<Client> {
...
+	public async pin(): Promise<Message> {
+		return this.channel.pins.add(this.id);
+	}
+
+	/**
+	 * Unpins the message to the channel
+	 * @since 0.0.4
+	 */
+	public async unpin(): Promise<ChannelPinsStore> {
+		return this.channel.pins.remove(this.id);
```
Comment (bdistin) [inline src/lib/caching/structures/messages/Message.ts:295]: Do we want to check if this.pinned at other devs?

File: src/lib/caching/structures/channels/DMChannel.ts [hunk 4]
```diff
@@ -58,6 +65,9 @@ export class DMChannel extends Channel {
 		super(client, data);
 		this.messages = new MessageStore(client, this);
 		this.typing = new Typing(this);
+
+		const filterMessagePins = this.messages.filter(msg => msg.pinned).keys();
```
Comment (bdistin) [inline src/lib/caching/structures/channels/DMChannel.ts:69]: At this point the MessageStore is empty.

File: src/lib/caching/stores/ChannelPinsStore.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	public async add(id: string, requestOptions: RequestOptions = {}): Promise<Message> {
+		const entry = await this.client.api.put(Routes.channelPin(this.channel.id, id), { ...requestOptions }) as APIMessageData;
+		this.set(entry.id);
+
+		// eslint-disable-next-line dot-notation
+		return this.channel.messages['_add'](entry);
+	}
+
+	/**
+	 * Deletes an invite given its code.
```
Comment (vladfrangu) [inline src/lib/caching/stores/ChannelPinsStore.ts:55]: invite eh?

File: src/lib/caching/stores/ChannelPinsStore.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	 * @param id The {@link Message#id message id} you want to pin
+	 * @param requestOptions The additional request options.
+	 * @see https://discord.com/developers/docs/resources/channel#add-pinned-channel-message
+	 */
+	public async add(id: string, requestOptions: RequestOptions = {}): Promise<Message> {
+		const entry = await this.client.api.put(Routes.channelPin(this.channel.id, id), { ...requestOptions }) as APIMessageData;
+		this.set(entry.id);
+
+		// eslint-disable-next-line dot-notation
+		return this.channel.messages['_add'](entry);
```
Comment (vladfrangu) [inline src/lib/caching/stores/ChannelPinsStore.ts:51]: Reading the docs would be a good idea, this entire block is wrong
image

File: src/lib/caching/structures/messages/Message.ts (hunk 3 above)
Comment (vladfrangu) [reply src/lib/caching/structures/messages/Message.ts:295]: Well, if we start doing certain checks for the end user, we might as well throw in permission checks in the lib before sending requests to Discord.. I say it's fine

File: src/lib/caching/structures/messages/Message.ts [hunk 7]
```diff
@@ -278,6 +279,22 @@ export class Message extends WebhookMessage<Client> {
...
+	 */
+	public async pin(): Promise<Message> {
+		return this.channel.pins.add(this.id);
+	}
+
+	/**
+	 * Unpins the message to the channel
+	 * @since 0.0.4
+	 */
+	public async unpin(): Promise<ChannelPinsStore> {
```
Comment (vladfrangu) [inline src/lib/caching/structures/messages/Message.ts:294]: This should return the message, just like `pin` :D

**Comment by Relms12345** [comment]
> This PR should also update our `CHANNEL_PINS_UPDATE` action to keep the pin stores up-to-date: [discord.com/developers/docs/topics/gateway#channel-pins-update](https://discord.com/developers/docs/topics/gateway#channel-pins-update)

I dont really know how to update actions so if someone can help out i will appreciate it

File: src/lib/caching/stores/ChannelPinsStore.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	/**
+	 * Removes a pin from the channel given the message ID.
+	 * @since 0.0.4
+	 * @param id The {@link Message#id message id}.
+	 * @param requestOptions The additional request options.
+	 * @see https://discord.com/developers/docs/resources/channel#delete-pinned-channel-message
+	 */
+	public async remove(id: string, requestOptions: RequestOptions = {}): Promise<this> {
+		await this.client.api.delete(Routes.channelPin(this.channel.id, id), requestOptions);
+		return this;
```
Comment (vladfrangu) [inline src/lib/caching/stores/ChannelPinsStore.ts:61]: remove doesn't do the opposite of add with `this.set`?

File: src/lib/caching/stores/ChannelPinsStore.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	/**
+	 * Returns a list of {@link Message messages}s with their metadata.
```
Comment (vladfrangu) [inline src/lib/caching/stores/ChannelPinsStore.ts:65]:
```suggestion
	 * Returns a list of {@link Message pinned messages}s with their metadata.
```

File: src/lib/caching/stores/ChannelPinsStore.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	 * Pins a message to the channel.
+	 * @since 0.0.4
+	 * @param id The {@link Message#id message id} you want to pin
+	 * @param requestOptions The additional request options.
+	 * @see https://discord.com/developers/docs/resources/channel#add-pinned-channel-message
+	 */
+	public async add(id: string, requestOptions: RequestOptions = {}): Promise<void> {
+		await this.client.api.put(Routes.channelPin(this.channel.id, id), { ...requestOptions });
+		this.set(id);
+		return;
```
Comment (vladfrangu) [inline src/lib/caching/stores/ChannelPinsStore.ts:49]: Why does add return void but remove returns this? Hell, why don't they both return this OR return the pinned message?

**Comment by vladfrangu** [comment]
> This PR should also update our `CHANNEL_PINS_UPDATE` action to keep the pin stores up-to-date: [discord.com/developers/docs/topics/gateway#channel-pins-update](https://discord.com/developers/docs/topics/gateway#channel-pins-update)

The gateway event only tells us something changed, not WHAT changed. We'd have to fetch the pins in order to keep the store up to date

File: src/lib/caching/structures/messages/Message.ts [hunk 11]
```diff
@@ -312,7 +312,13 @@ export class Message extends WebhookMessage<Client> {
 		if (data.attachments) for (const attachment of data.attachments) this.attachments.set(attachment.id, new MessageAttachment(attachment));
 		if (data.embeds) for (const embed of data.embeds) this.embeds.push(new Embed(embed));
 
-		if (isSet(data, 'pinned')) this.pinned = data.pinned;
+		if (isSet(data, 'pinned')) {
+			this.pinned = data.pinned;
+
+			// eslint-disable-next-line no-unused-expressions
+			this.pinned ? this.channel.pins.set(this.id) : this.channel.pins.delete(this.id);
```
Comment (kyranet) [inline src/lib/caching/structures/messages/Message.ts:319]:
```suggestion
			// eslint-disable-next-line no-unused-expressions
			if (this.pinned) this.channel.pins.set(this.id);
			else this.channel.pins.delete(this.id);
```

File: src/lib/caching/stores/ChannelPinsStore.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+	/**
+	 * Pins a message to the channel.
+	 * @since 0.0.4
+	 * @param id The {@link Message#id message id} you want to pin
+	 * @param requestOptions The additional request options.
+	 * @see https://discord.com/developers/docs/resources/channel#add-pinned-channel-message
+	 */
+	public async add(id: string, requestOptions: RequestOptions = {}): Promise<this> {
+		await this.client.api.put(Routes.channelPin(this.channel.id, id), { ...requestOptions });
```
Comment (kyranet) [inline src/lib/caching/stores/ChannelPinsStore.ts:47]:
```suggestion
		await this.client.api.put(Routes.channelPin(this.channel.id, id), requestOptions);
```

**Pull request merged**